### PR TITLE
[WIP] Fix for fyolo

### PR DESCRIPTION
**Pull request by primenumber**
Re-support fyolo_quantize_4_v4.

## Motivation and Context

Because of some modification of `tca_support`, dlk failed to compile fyolo_quantize_4_v4.
This PR re-support fyolo_quantize_4_v4.
But because of unknown reason, output compalison is failed now. → Output comparison matched!!!

## Description

- Fix layout propagation
- Fix quantized_conv2d
- Implement func_ExtractImagePatches for new layout
- Fix calculation threshold

## How has this been tested?

```shell
$ PYTHONPATH=python/dlk python python/dlk/scripts/generate_project.py -i examples/object_detection/fyolo_quantize_4_v4/minimal_graph_with_shape.pb -o tmp/ -p object_detection -hq -ts
$ cd tmp/object_detection
$ make -j8 lm_fpga
```

## Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature / Optimization (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist:

- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

**Comment by primenumber** [comment]
Input of 2nd binary convolution `block_3_conv_Conv2D` is correct.
But output of the convolution is wrong.
Output of lsb of specific channel is wrong.

**Comment by primenumber** [comment]
@n-nez @antonionevado Do you have any idea?

**Comment by antonionevado** [comment]
@primenumber 
A possible issue could be the thresholds values. Are you able to convert other models with these modifications, for example, widerface_v5?

**Comment by primenumber** [comment]
@antonionevado Yes, I'm able to convert widerface_v5 and output comparison is OK.

**Comment by antonionevado** [comment]
@primenumber 
I have been checking and this should fix your issue.
Change
```
        for c in range(ch):
            threshold_table[c, -1] = 1 \
                if np.all(threshold_table[c, 1:-1] > threshold_table[c, :-2], axis=0) else -1
            if np.all(threshold_table[c, 1:-1] == threshold_table[c, :-2], axis=0):
                threshold_table[c, -1] = 1
                threshold_table[c, 1:-1] = max_th_value

        # Put the thresholds into list
        conv_node.thresholds = threshold_table.flatten().tolist()
```
to:
```
        for c in range(ch):
            threshold_table[c, -1] = 1 \
                if np.all(threshold_table[c, 1:-1] > threshold_table[c, :-2], axis=0) else -1
            if np.all(threshold_table[c, 1:-1] == threshold_table[c, :-2], axis=0):
                threshold_table[c, -1] = 1
                threshold_table[c, 0:-1] = max_th_value

        # Put the thresholds into list
        conv_node.thresholds = threshold_table.flatten().tolist()
```

**Comment by primenumber** [comment]
@antonionevado Thank you!
Output comparison matched!!!

**Comment by primenumber** [comment]
@antonionevado @n-nez Please review.

File: dlk/python/dlk/templates/include/func/extract_image_patches.h [hunk 1]
```diff
@@ -23,6 +23,61 @@ limitations under the License.
...
+        for(T_UINT wj = 0; wj < out_width; wj++)
+          for(T_UINT ki = 0; ki < kernel_size; ki++)
+            for(T_UINT kj = 0; kj < kernel_size; kj++)
+            {
+              T_INT row = (wi * stride) + ki;
+              T_INT col = (wj * stride) + kj;
+              for(T_UINT digit = 0; digit < bits_per_input; ++digit) {
+                output(wi, wj,
+                  ih + (ki * kernel_size + kj) * input_depth,
+                  digit, 0) = input(row, col, ih, digit, 0);
```
Comment (n-nez) [inline dlk/python/dlk/templates/include/func/extract_image_patches.h]: Could we introduce additional variable here, and reformat it a bit, to make it a bit easier to comprehend?

E.g
```cpp
                T_UINT ch_high = ih + (ki * kernel_size + kj) * input_depth;
                output(wi, wj, ch_high, digit, 0) = input(row, col, ih, digit, 0);
```

File: dlk/python/dlk/templates/include/func/quantized_conv2d.h [hunk 2]
```diff
@@ -232,15 +232,15 @@ void func_QuantizedConv2D(
...
+          tmp_output[tmp_index++] = p.device_output_buf[h * (tca_channels * ncp.output_width) + w * tca_channels + d];
     Measurement::Stop();
 
     Measurement::Start("QuantizedConv2D_ApplyScalingFactor");
     int out_index = 0;
     for (int h = 0; h < ncp.output_height; ++h)
       for (int w = 0; w < ncp.output_width; ++w)
         for (int d = 0; d < ncp.output_channels; ++d)
-          output.data()[out_index++] = (scaling_factor[d] * post_qtz_factor) * p.device_output_buf[h * (tca_channels * ncp.input_width) + w * tca_channels + d];
+          output.data()[out_index++] = (scaling_factor[d] * post_qtz_factor) * p.device_output_buf[h * (tca_channels * ncp.output_width) + w * tca_channels + d];
```
Comment (n-nez) [inline dlk/python/dlk/templates/include/func/quantized_conv2d.h]: Special thanks for replacing `input_xxx` with `output_xxx` :+1: :sweat_smile:

File: dlk/python/dlk/templates/include/func/extract_image_patches.h (hunk 1 above)
Comment (primenumber) [reply dlk/python/dlk/templates/include/func/extract_image_patches.h]: Fixed.

Comment (n-nez) [reply dlk/python/dlk/templates/include/func/extract_image_patches.h]: Thank you!

**Pull request merged**